nav2_core/include/nav2_core/global_planner.hpp

```diff
@@ -0,0 +1,103 @@
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NAV2_CORE_GLOBAL_PLANNER_H_
+#define NAV2_CORE_GLOBAL_PLANNER_H_
+
+#include <string>
+#include "rclcpp/rclcpp.h"
+#include "nav2_costmap_2d/nav2_costmap_2d.hpp"
+#include "tf2_ros/Buffer.h"
+#include "nav2_msgs/msg/path.h"
+#include "geometry_msgs/msg/pose_stamped.h"
+
+namespace nav2_core
+{
+
+/**
+ * @class GlobalPlanner
+ * @brief Abstract interface for global planners to adhere to with pluginlib
+ */
+class GlobalPlanner
+{
+public:
+  /**
+   * @brief Virtual destructor
+   */
+  virtual ~GlobalPlanner() {}
+
+  /**
+   * @param  parent pointer to user's node
+   * @param  name The name of this planner
+   * @param  tf A pointer to a TF buffer
+   * @param  costmap_ros A pointer to the costmap
+   */
+  virtual void configure(const rclcpp::Node * parent,
+    const std::string & name, tf2_ros::Buffer * tf,
+    nav2_costmap_2d::Costmap2DROS * costmap_ros) = 0;
+
+  /**
+   * @brief Method to cleanup resources used on shutdown.
+   */
+  virtual void cleanup() = 0;
+
+  /**
+   * @brief Method to active planner and any threads involved in execution.
+   */
+  virtual void activate() = 0;
+
+  /**
+   * @brief Method to deactive planner and any threads involved in execution.
+   */
+  virtual void deactivate() = 0;
+
+  /**
+   * @brief Method to shutdown planner and any threads involved in execution.
+   */
+  virtual void shutdown() = 0;
+
+  /**
+   * @brief Method create the plan from a starting and ending goal.
+   * @param start The starting pose of the robot
+   * @param goal  The goal pose of the robot
+   * @return      The sequence of poses to get from start to goal, if any
+   */
+  virtual nav2_msgs::Path createPlan(
```

Comment:
I like having the method return the plan. Are you thinking throwing an exception if it can't compute a plan? Alternatively we could return a tuple.

Reply:
Potentially, though before we introduce exceptions here, I'd go to a struct return type with the type of return code and have the implicit path return. I really would not like to have exceptions anywhere they aren't totally necessary.

Reply:
After more thought, yes I think this is a place exceptions make sense. That way some custom planner can have custom exceptions deriving from some base exception class we provide.

Then those custom exceptions can be returned to the BT and have some custom recovery action based on it. This is a case where I think exceptions do make sense since we have to transport some unknown contextualized failure mode across a couple different places. 

EI failed -> exception thrown -> caught by planner server -> text of exception given to BT in failure update -> text can be reviewed by a BT condition node -> enact a specific recovery